firefox-ios/EcosiaTests/Account/Auth/AuthNotificationSystemTests.swift

```diff
@@ -0,0 +1,460 @@
+        
+        let action = AuthStateAction(
+            type: .userLoggedIn,
+            windowUUID: testWindowUUID,
+            isLoggedIn: true
+        )
+
+        // Act
+        authStateManager.dispatch(action: action, for: testWindowUUID)
+        
+        // Unsubscribe
+        authStateManager.unsubscribe(observer: observer)
+        
+        // Dispatch another action
+        let action2 = AuthStateAction(
+            type: .userLoggedOut,
+            windowUUID: testWindowUUID,
+            isLoggedIn: false
+        )
+        authStateManager.dispatch(action: action2, for: testWindowUUID)
+
+        // Assert
+        // Give some time for the notifications to be processed
+        let expectation = expectation(description: "Observer should receive limited notifications")
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: 1.0)
+        
+        XCTAssertEqual(observer.receivedNotifications.count, 1, "Observer should receive only one notification (before unsubscribe)")
+    }
+
+    // MARK: - Legacy Notification Compatibility Tests
+
+    func testLegacyNotificationNames_areAccessible() {
+        // Test that legacy notification names are accessible
+        let loginNotification = Notification.Name.EcosiaAuthDidLoginWithSessionToken
+        let logoutNotification = Notification.Name.EcosiaAuthDidLogout
+        let stateReadyNotification = Notification.Name.EcosiaAuthStateReady
+        let shouldLogoutFromWebNotification = Notification.Name.EcosiaAuthShouldLogoutFromWeb
+        
+        XCTAssertNotNil(loginNotification, "Login notification should be accessible")
+        XCTAssertNotNil(logoutNotification, "Logout notification should be accessible")
+        XCTAssertNotNil(stateReadyNotification, "State ready notification should be accessible")
+        XCTAssertNotNil(shouldLogoutFromWebNotification, "Should logout from web notification should be accessible")
+    }
+
+    func testLegacyNotificationObservation_worksWithNewSystem() {
+        // Arrange
+        notificationObserver.expectNotification(
+            name: .EcosiaAuthDidLoginWithSessionToken,
+            object: nil,
+            expectedCount: 1
+        )
+        
+        // Act - Post a legacy notification manually (simulating Auth.swift posting it)
+        NotificationCenter.default.post(name: .EcosiaAuthDidLoginWithSessionToken, object: nil)
+
+        // Assert
+        notificationObserver.waitForExpectations(timeout: 1.0)
+        
+        let receivedNotification = notificationObserver.receivedNotifications.first
+        XCTAssertNotNil(receivedNotification, "Should receive legacy notification")
+        XCTAssertEqual(receivedNotification?.name, .EcosiaAuthDidLoginWithSessionToken, "Should have correct notification name")
+    }
+
+    // MARK: - Notification Performance Tests
```

Comment:
Interesting idea for performance! Just wondering, what exactly could we do about low performance here? Isn't this just reflection iOS' internal observer library? If so, just wondering if we win much with this.

Reply:
These performance tests are more "nice to have" than "actionable". They would only be useful if:
- We were considering replacing NotificationCenter with an even more customized system (unlikely - not even mentioning Combine)
- We wanted to catch regressions in our userInfo payload size

I'm thinking of turning this into a regression test rather than "performance" - even those "many actions" I'm going to write are less likely to appear all at once.